DigitalLearningSolutions.Data/Services/LoginService.cs

```diff
@@ -0,0 +1,56 @@
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.Exceptions;
+    using DigitalLearningSolutions.Data.Models.User;
+
+    public interface ILoginService
+    {
+        (AdminUser?, DelegateUser?) VerifyUserDetailsAndGetClaims(string username, string password);
+    }
+
+    public class LoginService : ILoginService
+    {
+        private readonly ICryptoService cryptoService;
+        private readonly IUserService userService;
+
+        public LoginService(IUserService userService, ICryptoService cryptoService)
+        {
+            this.userService = userService;
+            this.cryptoService = cryptoService;
+        }
+
+        public (AdminUser?, DelegateUser?) VerifyUserDetailsAndGetClaims(string username, string password)
+        {
+            List<AdminUser> adminUsers = userService.GetAdminUsersByUsername(username);
+            List<DelegateUser> delegateUsers = userService.GetDelegateUsersByUsername(username);
+
+            if (adminUsers.Count == 0 && delegateUsers.Count == 0)
+            {
+                throw new UserAccountNotFoundException("No account with that email address or user ID could be found.");
```

Comment:
I think throwing exceptions to control the system flow is generally considered an anti-pattern. I'm wondering if we can split this into various functions and handle the cases of no user found, no password match and not approved on a controller level?

Reply:
Should be feasible enough. I think splitting it up into several functions was probably going to have be the first thing I was going to do in 361 - choose a centre.

Reply:
Actually, I think the entirety of the LoginService can just be moved into the controller instead. Pretty much all the LoginService is doing is grabbing some data from the UserService, and calling a CryptoService function, and doing some null checks on them. I'm not seeing anything that would be out of the ordinary in a controller, so I think I will get rid of the LoginService.

Reply:
I don't really agree here with the approach of pulling more logic into controllers, it makes things harder to unit test. Particularly as we now have a pretty significant chunk of logic around getting users and checking passwords here.